### ISSUE TITLE
Incorrect SQL for Assigned to Editor generated in User Collector

### ISSUE THREAD

**Issue by Vitaliy-1** [pkp/pkp-lib#7572]
**Describe the bug**
While debugging `assignedTo` property in Submission's Collector, I found that `Assigned To Editor` filter in Submission List Panel auto-suggests an incorrect user - in my example, the user that has a role as an author an a reviewer but not any editorial role.
Example of the request: 
```
GET /index.php/psp/api/v1/users?roleIds[]=16&roleIds[]=17&searchPhrase=Fra&_=1640357544860
```
Example of associated SQL:
```SQL
select `u`.* from `users` as `u` where exists 
(select * from `user_user_groups` as `uug` inner join `user_groups` as `ug` on `uug`.`user_group_id` = `ug`.`user_group_id` where `uug`.`user_id` = `u`.`user_id` and `ug`.`role_id` in (?, ?)) 
and 
LOWER(u.username) LIKE LOWER(?) or LOWER(u.email) LIKE LOWER(?) or exists (select * from `user_settings` as `us` where `us`.`user_id` = `u`.`user_id` and `us`.`setting_name` in (?, ?, ?, ?, ?, ?) and LOWER(us.setting_value) LIKE LOWER(?)) or exists (select * from `user_interests` as `ui` inner join `controlled_vocab_entry_settings` as `cves` on `ui`.`controlled_vocab_entry_id` = `cves`.`controlled_vocab_entry_id` where `ui`.`user_id` = `u`.`user_id` and LOWER(cves.setting_value) LIKE LOWER(?)) order by `u`.`user_id` asc
```
The relevant part here is that after subquery that retrieves users by role id:
```SQL
(select * from `user_user_groups` as `uug` inner join `user_groups` as `ug` on `uug`.`user_group_id` = `ug`.`user_group_id` where `uug`.`user_id` = `u`.`user_id` and `ug`.`role_id` in (?, ?)) 
```
There should be additional round brackets for the part that retrieves users by a keyword, like 
```SQL
and (LOWER(u.username) LIKE LOWER(?) or LOWER(u.email) LIKE LOWER(?)...) 
```

**What application are you using?**
 OJS, main branch

**Additional information**
At least for this case, User Collector's [search filter](https://github.com/pkp/pkp-lib/blob/38b0064f8f8164dff06ef085cadeeb25936c4d64/classes/user/Collector.inc.php#L535) should be put in a Laravel's `where` subquery

**Comment by Vitaliy-1** [comment]
Moreover, if implemented correctly, there still can be a situation, when the user, having an editorial role, is assigned as a reviewer to a submission and the current implementation doesn't filter out such submission - this is more relevant for OMP with the internal review stage.

**Comment by asmecher** [comment]
@NateWr, what is your preference?
- We change the "Assigned To Editor" filter so that only editor results are matched, or
- We relabel it to "Assigned To..." and leave it so that reviewer matches are also included.

**Comment by NateWr** [comment]
I think we can leave the filter and label alone. We actively discourage the use of editors as reviewers, so this is a use case that we permit but actively design against. And once #4975 is implemented, this problem will go away. Stage and reviewer assignments will be separated and the `assignedTo` prop will only consider stage assignments.

**Comment by NateWr** [comment]
@Vitaliy-1 can you reproduce this in 3.3? If so, we should assign the original bug report to a 3.3 milestone.

**Comment by Vitaliy-1** [comment]
No, able to reproduce only with the main branch; underlying SQL in 3.3 also looks different.

**Comment by asmecher** [comment]
@NateWr, the issue is that the current "Assigned to Editor"... filter actually *does* search both editor and reviewer assignments. The `assignedTo` filter in 3.3.0's `PKPSubmissionQueryBuilder` searches both places, like 3.4.0's `lib/pkp/classes/submission/Collector.inc.php` does, so I don't think the change between 3.3.0 and 3.4.0 is a query construction issue.

**Comment by NateWr** [comment]
The `assignedTo` parameter in the [/submissions endpoint](https://docs.pkp.sfu.ca/dev/api/ojs/3.3#tag/Submissions) of the REST API deliberately gets assignments from both places. This is required for the "My Queue" submission list, until #4975 is implemented. This shouldn't be a data leak because only Managers and Admins can use the `assignedTo` parameter. If that's not the case, this is an authorization bug.

In terms of the "Assigned to Editor" filter in the submissions list, the only options available in the autosuggest field [should be editors](https://github.com/pkp/pkp-lib/blob/main/classes/components/listPanels/PKPSubmissionsListPanel.inc.php#L160). Yes, if that editor is also assigned to a review on submission A, then submission A will be returned as "assigned to editor".

We design against this for "dual assignments", where the same user is assigned as a reviewer and in an editorial role to the same submission, because we discourage this practice. But I guess it's also possible that an editorial user is assigned as a reviewer, but _not assigned_ to the submission in an editorial role. In such cases the submission would be returned in the filter, which would be confusing. The same is true for author assignments, because the `assignedTo` parameter only checks for stage assignments.

I see that this is incorrect. But I don't think it's a data leak, in terms of exposing a reviewer's identity, because only the Manager or Admin can use this filter. And I'm not entirely convinced it's likely to be causing lots of problems for Managers.

However, if we want to fix this before implementing #4975, then we'll need to add a new parameter to the API/collector. We could either add a new, temporary parameter to the backend submissions handler to get just editorial assignments, `assignedEditorTo`. Or we could add a new endpoint like `/_submissions/mine` that matched submissions with stage and review assignments in one list. Then we can adjust the `assignedTo` to only match against stage assignments. If we think the Editor vs Author assignment distinction is important too, then maybe we should split this out into more params, like `assignedEditors`, `assignedAuthors`, `assignedReviewers`, and `assignedAssistants`.

**Comment by Vitaliy-1** [comment]
Just to be more specific. The roles the user has:
![autosuggest_roles](https://user-images.githubusercontent.com/20665587/148941747-d20a49f8-cb37-4b2f-9224-be8f455589b5.png)

Autosuggest field:
![autosuggest_user_field](https://user-images.githubusercontent.com/20665587/148941782-4293ec1e-4106-4a07-9344-7f3041abce82.png)

One of the results of the search:
![autosuggest_user_results1](https://user-images.githubusercontent.com/20665587/148941880-bd13ce48-fa64-4675-8090-cea3a11d7ed9.png)

This is different from behaviour in the 3.3 branch, where the user isn't suggested in the field at all. The logic in the User Collector that was before (with left joins) looks different from the one implemented in the main branch for this specific case. 
(But the logic in the Submission Collector looks the same)

**Comment by asmecher** [comment]
At least part of this turned out to be a subtle problem with the way we were using arrow function syntax. The following:
```PHP
$query->where(fn () => $query->severalConditionsHere)
```
...*is not the same as*...
```PHP
$query->where(fn ($query) => $query->severalConditionsHere)
```
...because the `$query` object is taken from the outer scope in the first, and the closure in the second. When chaining `OR` conditions together it's dangerous to use the outer scope `$query` (the first example) because they may result in the `OR` being applied to an earlier `WHERE` rather than being groupd (`a OR b`) as desired.

So: When using arrow functions as shorthand, be sure to include the `$query` parameter in `fn ($query) => ...`!

@Vitaliy-1, I've committed this change to `main`, but could you verify whether it resolves the entire issue, or just part?

**Comment by NateWr** [comment]
Yikes! Good to know, thanks!

**Comment by NateWr** [comment]
@pmangahis has filed a nice issue about the Assigned to Editor filter. So let's keep this issue specifically for the user collector bug and discuss the best behaviour for the Assigned to Editor filter at https://github.com/pkp/pkp-lib/issues/7611.

**Comment by asmecher** [comment]
I'm going to call this resolved.
